### ISSUE TITLE
CPU memory leak when using torch.no_grad()

### ISSUE THREAD

**Issue by chienpingtsung** [pytorch/pytorch#45334]
## 🐛 Bug

If use `torch.no_grad()` block, the cpu memory will continually increase untill OOM kill happens.
But once remove the `no_grad`, everything would be all right.
I tried del loss or put the validation step into a function, but the memory leak still happens.
Is my code wrong? or a BUG?

## To Reproduce

1. Validation step using `torch.no_grad()`.
2. Using `CrossEntropyLoss()` as criterion.
3. CPU RAM continually increasing occurs a OOM kill by system(Ubuntu 18.04).

PyTorch 1.6

```
for epoch in range(10):
    net.train()
    # Good training.
    for data in trainloader:
        inputs, labels = data['images'], data['masks']

        for idx in range(0, len(inputs), 7):
            optimizer.zero_grad()

            outputs = net(inputs[idx:idx + 7])
            loss = criterion(outputs, labels[idx:idx + 7])
            loss.backward()
            optimizer.step()

    # Bad validation.
    net.eval()
    test_loss = 0.0
    test_times = 0
    for data in testloader:
        # !!!!!!!!👇
        with torch.no_grad():
            inputs, labels = data['images'], data['masks']

            for idx in range(0, len(inputs), 7):
                # or put no_grad here, leaking still happens.
                outputs = net(inputs[idx:idx + 7])
                loss = criterion(outputs, labels[idx:idx + 7])
                test_loss += loss.item()
                test_times += 1
    test_loss /= test_times
```

## Expected behavior

Normally valid without increasing CPU RAM.

## Environment

 - PyTorch Version (e.g., 1.0): 1.6
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): NaN
 - Python version: 3.8.5
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: Tesla V100 16G
 - Any other relevant information: with system RAM 128G

## Additional context

Thanks for all your excellent work!

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved

**Comment by zou3519** [comment]
hi-pri because memory leak seems serious. I have not attempted to repro this yet.

**Comment by albanD** [comment]
Cross post from the forum: https://discuss.pytorch.org/t/memory-leak-if-use-torch-no-grad/97484

We can continue the discussion to pinpoint the cause on the forum and re-open this if it actually is a bug on pytorch side.

**Comment by davidireland3** [comment]
I am having similar problems, I commented in the other discussion [here](https://discuss.pytorch.org/t/memory-leak-if-use-torch-no-grad/97484/5?u=david_ireland). Any help is appreciated.

**Comment by shaoyuyoung** [comment]
sry to bother

Is there any progress on this issue?
